File: haskell/Day01.hs
module Main where

import Common

-- Calculate the fuel for a list of masses
sumFuel :: [Int] -> Int
sumFuel list = sum (map calculateFuel list)

-- Calculate the extended fuel for a list of masses
sumExtendedFuel :: [Int] -> Int
sumExtendedFuel list = sum (map calculateExtendendFuel list)

-- Calculate the fuel by mass (and fuel mass,...)
calculateExtendendFuel :: Int -> Int
calculateExtendendFuel mass = if fuel <= 0 then 0 else fuel + (calculateExtendendFuel fuel)
  where fuel = calculateFuel mass

-- Calculate the fuel by mass
calculateFuel :: Int -> Int
calculateFuel mass = mass `div` 3 - 2

{- Handle input and solve executions -}

-- Solve first challenge
solve1 :: Input -> Int
solve1 (Lines (Just list)) = sumFuel $ map read list

-- Solve second challenge
solve2 :: Input -> Int
solve2 (Lines (Just list)) = sumExtendedFuel $ map read list

-- Print challenge result
main = do
  solve "Sum of fuel requirements" (Lines Nothing) (solve1)
  solve "Sum of extended fuel requirements" (Lines Nothing) (solve2)

File: clojure/Day01.clj
(ns Day01)

; Calculates the fuel for a given mass (m div 3 - 2)
(defn calc-fuel [mass]
  (- (quot mass 3) 2))

; Calculates the fuel for a given mass + the fuel for that fuel + the fuel for that fuel etc.. until 0 is reached.
; Returns the sum of all these.
(defn calc-recursive-fuel [mass]
  (apply + (take-while pos? (drop 1 (iterate calc-fuel mass)))))

; Returns the sum of the (simple) fuel values of the masses
(defn total-fuel-simple [masses]
  (apply + (map calc-fuel masses)))

; Returns the sum of the (recursive) fuel values of the masses
(defn total-fuel-recursive [masses]
  (apply + (map calc-recursive-fuel masses)))

; Reads and parses the lines of a file to ints. Returns them as a sequence (not lazy)
(defn read-ints [file-path]
  (with-open [reader (clojure.java.io/reader file-path)]
    (doall (map #(Integer/parseInt %) (line-seq reader)))))

(def input (read-ints (first *command-line-args*)))

(println "Fuel required:" (total-fuel-simple input))
(println "Fuel required (including fuel for fuel):" (total-fuel-recursive input))